linux/bindings/src/thread.rs

```diff
@@ -0,0 +1,266 @@
+// SPDX-License-Identifier: GPL-2.0
+// Copyright (c) 2023 Ant Group CO., Ltd.
+
+//! Kernel thread.
+
+use core::{
+    cell::Cell,
+    fmt,
+    marker::PhantomData,
+    ptr::NonNull,
+    sync::atomic::{AtomicBool, Ordering},
+    time::Duration,
+};
+use kernel::{
+    new_condvar, new_mutex,
+    prelude::*,
+    sync::{CondVar, Mutex},
+};
+
+/// A handle to kernel's `struct task_struct`.
+pub struct Thread {
+    inner: NonNull<crate::task_struct>,
+}
+
+/// An owned permission to join on a `Thread` (block on its termination).
+///
+/// This struct is created by the `spawn` function.
+pub struct JoinHandle<T> {
+    thread: Thread,
+    should_stop: AtomicBool,
+    _p: PhantomData<T>,
+}
+
+/// An item to hold the user closure and its return value for `Thread`.
+///
+/// The `closure` can only be accessed in the `threadfn` of one kthread,
+/// so it's safe to use `Cell`.
+#[pin_data]
+struct ThreadItem<T> {
+    closure: Cell<Option<Box<dyn FnOnce() -> T>>>,
+    #[pin]
+    value: Mutex<Option<T>>,
+    #[pin]
+    finished: CondVar,
+}
+
+impl<T> ThreadItem<T> {
+    /// Constructs an in-place fallible initializer.
+    fn try_new<F>(f: F) -> impl PinInit<Self, Error>
+    where
+        F: FnOnce() -> T + Send + 'static,
+        T: Send + 'static,
+    {
+        try_pin_init!(Self {
+            closure: {
+                let closure: Box<dyn FnOnce() -> T> = Box::try_new(f)?;
+                Cell::new(Some(closure))
+            },
+            value <- new_mutex!(None),
+            finished <- new_condvar!(),
+        })
+    }
+
+    /// Executes the user closure.
+    ///
+    /// # Panics
+    ///
+    /// This method must only be called once.
+    fn execute(&self) {
+        let closure = self.closure.take().expect("take closure more than once");
+        let mut value = self.value.lock();
+        *value = Some(closure());
```

Comment:
I am a bit worried about executing the closure while holding the lock. Is it possible that `ThreadItem::is_finished` or `ThreadItem::wait_value` is called during the period when the `ThreadItem::execute` is on-going? This would lead to deadlock.

Reply:
You're right. `ThreadItem::is_finished` should not be blocked while `ThreadItem::execute` is on-going. I'll fix it.